ledger/streaming/client/src/scp_validator.rs

```diff
@@ -0,0 +1,343 @@
+    /// Create new validator object
+    pub fn new(local_node_id: ID, local_quorum_set: QuorumSet<ID>, logger: Logger) -> Self {
+        Self {
+            local_node_id,
+            local_quorum_set,
+            slots_to_externalized_blocks: HashMap::new(),
+            highest_slot_index: 0,
+            logger,
+        }
+    }
+
+    ///Fill slot with received component for certain block_height
+    pub fn update_slot(&mut self, node_id: ID, component: BlockStreamComponents) {
+        let index = component.block_data.block().index;
+        if let std::collections::hash_map::Entry::Vacant(e) =
+            self.slots_to_externalized_blocks.entry(index)
+        {
+            let mut node_map = HashMap::new();
+            node_map.insert(node_id, component);
+            e.insert(node_map);
+        } else {
+            let node_map = self.slots_to_externalized_blocks.get_mut(&index).unwrap();
+            node_map.insert(node_id, component);
+        }
+    }
+
+    /// After block is externalized, remove records at previous slot
+    pub fn set_highest_externalized_slot(&mut self, index: SlotIndex) {
+        self.highest_slot_index = index;
+    }
+
+    /// Determine if we can externalize a block, if so return the block
+    pub fn attempt_externalize_block(&mut self) -> Option<BlockStreamComponents> {
+        let index = self.highest_slot_index + 1;
+
+        // If blocks received so far are less than quorum, don't proceed
+        if let Some(ballots) = self.slots_to_externalized_blocks.get(&index) {
+            if self.local_quorum_set.threshold > ballots.len() as u32 {
+                return None;
+            }
+        }
+
+        // Else check for a quorum and see if the blocks externalized consitute a
+        // quorum slice
+        let ballot_map = self.find_quorum(self.local_quorum_set.clone());
+        let block_id = self.get_quorum_slice_vote(self.local_quorum_set.threshold, &ballot_map);
+
+        // If there's consensus, return the block contents
+        if let Some(block_id) = block_id {
+            let winning_ballots = ballot_map.get(&block_id).unwrap();
+            for ballot in winning_ballots {
+                if let QuorumSetMember::Node(id) = ballot {
+                    let component = self.slots_to_externalized_blocks.get(&index).unwrap();
+                    self.highest_slot_index += 1;
+                    log::trace!(
+                        self.logger,
+                        "Node: {} externalizing block: {}",
+                        self.local_node_id,
+                        self.highest_slot_index,
+                    );
+
+                    return Some(component.get(id).unwrap().clone());
+                }
+            }
+        }
+
+        // If not let consumers know
+        None
+    }
+
+    /// Check to find there's a quorum on the value of the blocks
+    fn find_quorum(&self, quorum_set: QuorumSet<ID>) -> HashMap<BlockID, Vec<QuorumSetMember<ID>>> {
```

Comment:
The hard part is when you start recursing, you real quickly run into lifetime issues. It would be better not to clone so I'll take another try at it though.

Reply:
Got it done